### DISCUSSION TITLE
Make Factory recycle method to accept the name of the relationship

### DISCUSSION THREAD

**Discussion by khalyomede** [laravel/framework#46873] [Ideas]
## Context

I currently use factories to propagate my user across various models:

```php
public function testUserOnStandardPlanCannotTransferOwnershipWhenAtLeastOneOfItsAreaIsAttachedToAnItem(): void
{
  $user = User::factory()
    ->has(Subscription::factory()->standard())
    ->create();

  $shoppingList = ShoppingList::factory()
    ->has(User::factory()->has(Subscription::factory()->standard()), "author")
    ->has(User::factory(), "members")
    ->has(Item::factory()->for(Aisle::factory()->recycle($user)))
    ->create();
}
```

As you can see, I was able to save up some line of code in line 10

```php
->has(Item::factory()->for(Aisle::factory()->recycle($user)))
```

But I have failed to recycle the `$user` in line 8 for defining my shopping list author, because I need to define the relationship name

```php
->has(User::factory()->has(Subscription::factory()->standard()), "author")
```

## Proposal

Allow the `->recycle()` method to take the name of the relationship as second parameter, much like `->for()` allows it. This would help me not repeat my code above:

```php
public function testUserOnStandardPlanCannotTransferOwnershipWhenAtLeastOneOfItsAreaIsAttachedToAnItem(): void
{
  $user = User::factory()
    ->has(Subscription::factory()->standard())
    ->create();

  $shoppingList = ShoppingList::factory()
    ->recycle($user, "author")
    ->has(User::factory(), "members")
    ->has(Item::factory()->for(Aisle::factory()->recycle($user)))
    ->create();
}
```

What do you think about it?

**Comment by tylerwiegand** [comment]
I admit this would be helpful for me as well, but only because the original writers of the software chose to go against the grain, against convention, and did not use the convention dictated name for the relation. In my case, we have several different names for each relation to the same model -- which means I would need multiple names accepted to the recycle method in order to cover all the possibilities. This, to me, is us not being willing/able to refactor our own code rather than needing another feature from the Laravel team.

Embrace convention! =)
